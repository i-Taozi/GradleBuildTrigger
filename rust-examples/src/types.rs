///类型
pub fn types() {
    let decimal = 65.4321_f32;
    //不支持隐式转化
    //let integer: u8 = decimal;
    // FIXME ^ Comment out this line

    //显示转化
    let integer = decimal as u8;
    let character = integer as char;

    println!("Casting: {} -> {} -> {}", decimal, integer, character);

    // 1000 already fits in a u16
    println!("1000 as a u16 is: {}", 1000 as u16);

    // 1000 - 256 - 256 - 256 = 232
    // Under the hood, the first 8 least significant bits (LSB) are kept,
    // while the rest towards the most significant bit (MSB) get truncated.
    //println!("1000 as a u8 is : {}", 1000 as u8);
    // -1 + 256 = 255
    println!("  -1 as a u8 is : {}", (-1i8) as u8);
    println!("1000 mod 256 is : {}", 1000 % 256);

    println!(" 128 as a i16 is: {}", 128 as i16);
    //println!(" 128 as a i8 is : {}", 128 as i8);
    //println!("1000 as a u8 is : {}", 1000 as u8);
    //println!(" 232 as a i8 is : {}", 232 as i8);

    literals();
}

fn literals() {
    //类型作为后缀
    let x = 1u8;
    let y = 2u32;
    let z = 3f32;

    //非固定文字，其类型取决于如何使用
    let i = 1;
    let f = 1.0;

    //Returns the size of the pointed-to value in bytes.
    println!("size of `x` in bytes: {}", std::mem::size_of_val(&x));
    println!("size of `y` in bytes: {}", std::mem::size_of_val(&y));
    println!("size of `z` in bytes: {}", std::mem::size_of_val(&z));
    println!("size of `i` in bytes: {}", std::mem::size_of_val(&i));
    println!("size of `f` in bytes: {}", std::mem::size_of_val(&f));
}
