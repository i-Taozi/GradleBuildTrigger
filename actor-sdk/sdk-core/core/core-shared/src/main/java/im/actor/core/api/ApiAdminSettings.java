package im.actor.core.api;
/*
 *  Generated by the Actor API Scheme generator.  DO NOT EDIT!
 */

import im.actor.runtime.bser.*;
import im.actor.runtime.collections.*;
import static im.actor.runtime.bser.Utils.*;
import im.actor.core.network.parser.*;
import org.jetbrains.annotations.Nullable;
import org.jetbrains.annotations.NotNull;
import com.google.j2objc.annotations.ObjectiveCName;
import java.io.IOException;
import java.util.List;
import java.util.ArrayList;

public class ApiAdminSettings extends BserObject {

    private boolean showAdminsToMembers;
    private boolean canMembersInvite;
    private boolean canMembersEditGroupInfo;
    private boolean canAdminsEditGroupInfo;
    private boolean showJoinLeaveMessages;

    public ApiAdminSettings(boolean showAdminsToMembers, boolean canMembersInvite, boolean canMembersEditGroupInfo, boolean canAdminsEditGroupInfo, boolean showJoinLeaveMessages) {
        this.showAdminsToMembers = showAdminsToMembers;
        this.canMembersInvite = canMembersInvite;
        this.canMembersEditGroupInfo = canMembersEditGroupInfo;
        this.canAdminsEditGroupInfo = canAdminsEditGroupInfo;
        this.showJoinLeaveMessages = showJoinLeaveMessages;
    }

    public ApiAdminSettings() {

    }

    public boolean showAdminsToMembers() {
        return this.showAdminsToMembers;
    }

    public boolean canMembersInvite() {
        return this.canMembersInvite;
    }

    public boolean canMembersEditGroupInfo() {
        return this.canMembersEditGroupInfo;
    }

    public boolean canAdminsEditGroupInfo() {
        return this.canAdminsEditGroupInfo;
    }

    public boolean showJoinLeaveMessages() {
        return this.showJoinLeaveMessages;
    }

    @Override
    public void parse(BserValues values) throws IOException {
        this.showAdminsToMembers = values.getBool(1);
        this.canMembersInvite = values.getBool(2);
        this.canMembersEditGroupInfo = values.getBool(3);
        this.canAdminsEditGroupInfo = values.getBool(4);
        this.showJoinLeaveMessages = values.getBool(5);
        if (values.hasRemaining()) {
            setUnmappedObjects(values.buildRemaining());
        }
    }

    @Override
    public void serialize(BserWriter writer) throws IOException {
        writer.writeBool(1, this.showAdminsToMembers);
        writer.writeBool(2, this.canMembersInvite);
        writer.writeBool(3, this.canMembersEditGroupInfo);
        writer.writeBool(4, this.canAdminsEditGroupInfo);
        writer.writeBool(5, this.showJoinLeaveMessages);
        if (this.getUnmappedObjects() != null) {
            SparseArray<Object> unmapped = this.getUnmappedObjects();
            for (int i = 0; i < unmapped.size(); i++) {
                int key = unmapped.keyAt(i);
                writer.writeUnmapped(key, unmapped.get(key));
            }
        }
    }

    @Override
    public String toString() {
        String res = "struct AdminSettings{";
        res += "showAdminsToMembers=" + this.showAdminsToMembers;
        res += ", canMembersInvite=" + this.canMembersInvite;
        res += ", canMembersEditGroupInfo=" + this.canMembersEditGroupInfo;
        res += ", canAdminsEditGroupInfo=" + this.canAdminsEditGroupInfo;
        res += ", showJoinLeaveMessages=" + this.showJoinLeaveMessages;
        res += "}";
        return res;
    }

}
