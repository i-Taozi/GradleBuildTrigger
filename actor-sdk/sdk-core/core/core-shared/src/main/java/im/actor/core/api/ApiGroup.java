package im.actor.core.api;
/*
 *  Generated by the Actor API Scheme generator.  DO NOT EDIT!
 */

import im.actor.runtime.bser.*;
import im.actor.runtime.collections.*;
import static im.actor.runtime.bser.Utils.*;
import im.actor.core.network.parser.*;
import org.jetbrains.annotations.Nullable;
import org.jetbrains.annotations.NotNull;
import com.google.j2objc.annotations.ObjectiveCName;
import java.io.IOException;
import java.util.List;
import java.util.ArrayList;

public class ApiGroup extends BserObject {

    private int id;
    private long accessHash;
    private String title;
    private ApiAvatar avatar;
    private Integer membersCount;
    private Boolean isMember;
    private Boolean isHidden;
    private ApiGroupType groupType;
    private Long permissions;
    private Boolean isDeleted;
    private ApiMapValue ext;

    public ApiGroup(int id, long accessHash, @NotNull String title, @Nullable ApiAvatar avatar, @Nullable Integer membersCount, @Nullable Boolean isMember, @Nullable Boolean isHidden, @Nullable ApiGroupType groupType, @Nullable Long permissions, @Nullable Boolean isDeleted, @Nullable ApiMapValue ext) {
        this.id = id;
        this.accessHash = accessHash;
        this.title = title;
        this.avatar = avatar;
        this.membersCount = membersCount;
        this.isMember = isMember;
        this.isHidden = isHidden;
        this.groupType = groupType;
        this.permissions = permissions;
        this.isDeleted = isDeleted;
        this.ext = ext;
    }

    public ApiGroup() {

    }

    public int getId() {
        return this.id;
    }

    public long getAccessHash() {
        return this.accessHash;
    }

    @NotNull
    public String getTitle() {
        return this.title;
    }

    @Nullable
    public ApiAvatar getAvatar() {
        return this.avatar;
    }

    @Nullable
    public Integer getMembersCount() {
        return this.membersCount;
    }

    @Nullable
    public Boolean isMember() {
        return this.isMember;
    }

    @Nullable
    public Boolean isHidden() {
        return this.isHidden;
    }

    @Nullable
    public ApiGroupType getGroupType() {
        return this.groupType;
    }

    @Nullable
    public Long getPermissions() {
        return this.permissions;
    }

    @Nullable
    public Boolean isDeleted() {
        return this.isDeleted;
    }

    @Nullable
    public ApiMapValue getExt() {
        return this.ext;
    }

    @Override
    public void parse(BserValues values) throws IOException {
        this.id = values.getInt(1);
        this.accessHash = values.getLong(2);
        this.title = values.getString(3);
        this.avatar = values.optObj(4, new ApiAvatar());
        this.membersCount = values.optInt(24);
        this.isMember = values.optBool(6);
        this.isHidden = values.optBool(20);
        int val_groupType = values.getInt(25, 0);
        if (val_groupType != 0) {
            this.groupType = ApiGroupType.parse(val_groupType);
        }
        this.permissions = values.optLong(26);
        this.isDeleted = values.optBool(27);
        this.ext = values.optObj(22, new ApiMapValue());
        if (values.hasRemaining()) {
            setUnmappedObjects(values.buildRemaining());
        }
    }

    @Override
    public void serialize(BserWriter writer) throws IOException {
        writer.writeInt(1, this.id);
        writer.writeLong(2, this.accessHash);
        if (this.title == null) {
            throw new IOException();
        }
        writer.writeString(3, this.title);
        if (this.avatar != null) {
            writer.writeObject(4, this.avatar);
        }
        if (this.membersCount != null) {
            writer.writeInt(24, this.membersCount);
        }
        if (this.isMember != null) {
            writer.writeBool(6, this.isMember);
        }
        if (this.isHidden != null) {
            writer.writeBool(20, this.isHidden);
        }
        if (this.groupType != null) {
            writer.writeInt(25, this.groupType.getValue());
        }
        if (this.permissions != null) {
            writer.writeLong(26, this.permissions);
        }
        if (this.isDeleted != null) {
            writer.writeBool(27, this.isDeleted);
        }
        if (this.ext != null) {
            writer.writeObject(22, this.ext);
        }
        if (this.getUnmappedObjects() != null) {
            SparseArray<Object> unmapped = this.getUnmappedObjects();
            for (int i = 0; i < unmapped.size(); i++) {
                int key = unmapped.keyAt(i);
                writer.writeUnmapped(key, unmapped.get(key));
            }
        }
    }

    @Override
    public String toString() {
        String res = "struct Group{";
        res += "id=" + this.id;
        res += ", title=" + this.title;
        res += ", avatar=" + (this.avatar != null ? "set":"empty");
        res += ", membersCount=" + this.membersCount;
        res += ", isMember=" + this.isMember;
        res += ", isHidden=" + this.isHidden;
        res += ", groupType=" + this.groupType;
        res += ", permissions=" + this.permissions;
        res += ", isDeleted=" + this.isDeleted;
        res += ", ext=" + this.ext;
        res += "}";
        return res;
    }

}
