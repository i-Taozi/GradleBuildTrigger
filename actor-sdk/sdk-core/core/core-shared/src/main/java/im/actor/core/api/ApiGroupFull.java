package im.actor.core.api;
/*
 *  Generated by the Actor API Scheme generator.  DO NOT EDIT!
 */

import im.actor.runtime.bser.*;
import im.actor.runtime.collections.*;
import static im.actor.runtime.bser.Utils.*;
import im.actor.core.network.parser.*;
import org.jetbrains.annotations.Nullable;
import org.jetbrains.annotations.NotNull;
import com.google.j2objc.annotations.ObjectiveCName;
import java.io.IOException;
import java.util.List;
import java.util.ArrayList;

public class ApiGroupFull extends BserObject {

    private int id;
    private long createDate;
    private Integer ownerUid;
    private List<ApiMember> members;
    private String theme;
    private String about;
    private ApiMapValue ext;
    private Boolean isAsyncMembers;
    private Boolean isSharedHistory;
    private String shortName;
    private Long permissions;

    public ApiGroupFull(int id, long createDate, @Nullable Integer ownerUid, @NotNull List<ApiMember> members, @Nullable String theme, @Nullable String about, @Nullable ApiMapValue ext, @Nullable Boolean isAsyncMembers, @Nullable Boolean isSharedHistory, @Nullable String shortName, @Nullable Long permissions) {
        this.id = id;
        this.createDate = createDate;
        this.ownerUid = ownerUid;
        this.members = members;
        this.theme = theme;
        this.about = about;
        this.ext = ext;
        this.isAsyncMembers = isAsyncMembers;
        this.isSharedHistory = isSharedHistory;
        this.shortName = shortName;
        this.permissions = permissions;
    }

    public ApiGroupFull() {

    }

    public int getId() {
        return this.id;
    }

    public long getCreateDate() {
        return this.createDate;
    }

    @Nullable
    public Integer getOwnerUid() {
        return this.ownerUid;
    }

    @NotNull
    public List<ApiMember> getMembers() {
        return this.members;
    }

    @Nullable
    public String getTheme() {
        return this.theme;
    }

    @Nullable
    public String getAbout() {
        return this.about;
    }

    @Nullable
    public ApiMapValue getExt() {
        return this.ext;
    }

    @Nullable
    public Boolean isAsyncMembers() {
        return this.isAsyncMembers;
    }

    @Nullable
    public Boolean isSharedHistory() {
        return this.isSharedHistory;
    }

    @Nullable
    public String getShortName() {
        return this.shortName;
    }

    @Nullable
    public Long getPermissions() {
        return this.permissions;
    }

    @Override
    public void parse(BserValues values) throws IOException {
        this.id = values.getInt(1);
        this.createDate = values.getLong(6);
        this.ownerUid = values.optInt(5);
        List<ApiMember> _members = new ArrayList<ApiMember>();
        for (int i = 0; i < values.getRepeatedCount(12); i ++) {
            _members.add(new ApiMember());
        }
        this.members = values.getRepeatedObj(12, _members);
        this.theme = values.optString(2);
        this.about = values.optString(3);
        this.ext = values.optObj(7, new ApiMapValue());
        this.isAsyncMembers = values.optBool(11);
        this.isSharedHistory = values.optBool(10);
        this.shortName = values.optString(14);
        this.permissions = values.optLong(27);
        if (values.hasRemaining()) {
            setUnmappedObjects(values.buildRemaining());
        }
    }

    @Override
    public void serialize(BserWriter writer) throws IOException {
        writer.writeInt(1, this.id);
        writer.writeLong(6, this.createDate);
        if (this.ownerUid != null) {
            writer.writeInt(5, this.ownerUid);
        }
        writer.writeRepeatedObj(12, this.members);
        if (this.theme != null) {
            writer.writeString(2, this.theme);
        }
        if (this.about != null) {
            writer.writeString(3, this.about);
        }
        if (this.ext != null) {
            writer.writeObject(7, this.ext);
        }
        if (this.isAsyncMembers != null) {
            writer.writeBool(11, this.isAsyncMembers);
        }
        if (this.isSharedHistory != null) {
            writer.writeBool(10, this.isSharedHistory);
        }
        if (this.shortName != null) {
            writer.writeString(14, this.shortName);
        }
        if (this.permissions != null) {
            writer.writeLong(27, this.permissions);
        }
        if (this.getUnmappedObjects() != null) {
            SparseArray<Object> unmapped = this.getUnmappedObjects();
            for (int i = 0; i < unmapped.size(); i++) {
                int key = unmapped.keyAt(i);
                writer.writeUnmapped(key, unmapped.get(key));
            }
        }
    }

    @Override
    public String toString() {
        String res = "struct GroupFull{";
        res += "id=" + this.id;
        res += ", createDate=" + this.createDate;
        res += ", ownerUid=" + this.ownerUid;
        res += ", members=" + this.members;
        res += ", theme=" + this.theme;
        res += ", about=" + this.about;
        res += ", isAsyncMembers=" + this.isAsyncMembers;
        res += ", isSharedHistory=" + this.isSharedHistory;
        res += ", shortName=" + this.shortName;
        res += ", permissions=" + this.permissions;
        res += "}";
        return res;
    }

}
