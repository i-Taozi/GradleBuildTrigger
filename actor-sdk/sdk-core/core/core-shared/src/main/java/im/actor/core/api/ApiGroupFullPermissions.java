package im.actor.core.api;
/*
 *  Generated by the Actor API Scheme generator.  DO NOT EDIT!
 */

import java.io.IOException;

public enum ApiGroupFullPermissions {

    EDIT_INFO(1),
    VIEW_MEMBERS(2),
    INVITE_MEMBERS(3),
    INVITE_VIA_LINK(4),
    CALL(5),
    EDIT_ADMIN_SETTINGS(6),
    VIEW_ADMINS(7),
    EDIT_ADMINS(8),
    KICK_INVITED(9),
    KICK_ANYONE(10),
    EDIT_FOREIGN(11),
    DELETE_FOREIGN(12),
    UNSUPPORTED_VALUE(-1);

    private int value;

    ApiGroupFullPermissions(int value) {
        this.value = value;
    }

    public int getValue() {
        return value;
    }

    public static ApiGroupFullPermissions parse(int value) throws IOException {
        switch(value) {
            case 1: return ApiGroupFullPermissions.EDIT_INFO;
            case 2: return ApiGroupFullPermissions.VIEW_MEMBERS;
            case 3: return ApiGroupFullPermissions.INVITE_MEMBERS;
            case 4: return ApiGroupFullPermissions.INVITE_VIA_LINK;
            case 5: return ApiGroupFullPermissions.CALL;
            case 6: return ApiGroupFullPermissions.EDIT_ADMIN_SETTINGS;
            case 7: return ApiGroupFullPermissions.VIEW_ADMINS;
            case 8: return ApiGroupFullPermissions.EDIT_ADMINS;
            case 9: return ApiGroupFullPermissions.KICK_INVITED;
            case 10: return ApiGroupFullPermissions.KICK_ANYONE;
            case 11: return ApiGroupFullPermissions.EDIT_FOREIGN;
            case 12: return ApiGroupFullPermissions.DELETE_FOREIGN;
            default: return ApiGroupFullPermissions.UNSUPPORTED_VALUE;
        }
    }
}
