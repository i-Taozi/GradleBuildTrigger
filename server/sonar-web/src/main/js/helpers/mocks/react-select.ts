/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import { GroupTypeBase, OptionProps } from 'react-select';

export function mockReactSelectOptionProps<
  OptionType,
  IsMulti extends boolean,
  GroupType extends GroupTypeBase<OptionType> = GroupTypeBase<OptionType>
>(
  data: OptionType,
  overrides?: OptionProps<OptionType, IsMulti, GroupType>
): OptionProps<OptionType, IsMulti, GroupType> {
  return {
    ...overrides,
    data
  } as OptionProps<OptionType, IsMulti, GroupType>;
}
