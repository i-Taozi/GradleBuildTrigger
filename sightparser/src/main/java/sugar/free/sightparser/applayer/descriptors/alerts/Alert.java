package sugar.free.sightparser.applayer.descriptors.alerts;

import java.io.Serializable;
import java.util.HashMap;
import java.util.Map;

import sugar.free.sightparser.pipeline.ByteBuf;

public abstract class Alert implements Serializable {

    private static final long serialVersionUID = 1L;

    public static final Map<Short, Class<? extends Alert>> ALERTS = new HashMap<>();

    static {
        ALERTS.put((short) 0x1F00, Reminder1DeliverBolus.class);
        ALERTS.put((short) 0xE300, Reminder2MissedBolus.class);
        ALERTS.put((short) 0xFC00, Reminder3AlarmClock.class);
        ALERTS.put((short) 0x2503, Reminder4ChangeInfusionSet.class);
        ALERTS.put((short) 0x3A03, Reminder7TBRCompleted.class);

        ALERTS.put((short) 0xC603, Warning31CartridgeLow.class);
        ALERTS.put((short) 0xD903, Warning32BatteryLow.class);
        ALERTS.put((short) 0x4A05, Warning33InvalidDateTime.class);
        ALERTS.put((short) 0x5505, Warning34EndOfWarranty.class);
        ALERTS.put((short) 0xA905, Warning36TBRCancelled.class);
        ALERTS.put((short) 0xB605, Warning38BolusCancelled.class);
        ALERTS.put((short) 0x6F06, Warning39LoantimeWarning.class);

        ALERTS.put((short) 0x7006, Maintenance20CartridgeNotInserted.class);
        ALERTS.put((short) 0x8C06, Maintenance21CartridgeEmpty.class);
        ALERTS.put((short) 0x9306, Maintenance22BatteryEmpty.class);
        ALERTS.put((short) 0x2618, Maintenance23AutomaticOff.class);
        ALERTS.put((short) 0x3918, Maintenance24Occlusion.class);
        ALERTS.put((short) 0xC518, Maintenance25LoantimeOver.class);
        ALERTS.put((short) 0xDA18, Maintenance26CartridgeChangeNotCompleted.class);
        ALERTS.put((short) 0x031B, Maintenance27DataDownloadFailed.class);
        ALERTS.put((short) 0x1C1B, Maintenance28PauseModeTimeout.class);
        ALERTS.put((short) 0xE01B, Maintenance29BatteryTypeNotSet.class);
        ALERTS.put((short) 0xFF1B, Maintenance30CartridgeTypeNotSet.class);

        ALERTS.put((short) 0x6C1D, Error6MechanicalError.class);
        ALERTS.put((short) 0x731D, Error10RewindError.class);
        ALERTS.put((short) 0x8F1D, Error13LanguageError.class);
    }

    public void parse(ByteBuf byteBuf) {
    }

}
