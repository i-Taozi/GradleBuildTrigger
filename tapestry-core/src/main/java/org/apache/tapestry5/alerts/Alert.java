// Copyright 2011 The Apache Software Foundation
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package org.apache.tapestry5.alerts;

import org.apache.tapestry5.ioc.internal.util.InternalUtils;
import org.apache.tapestry5.json.JSONObject;

import java.io.Serializable;
import java.util.concurrent.atomic.AtomicLong;

/**
 * An Alert that may be presented to the user. The Alert has a message, but also includes
 * a severity (that controls how it is presented to the user), and a duration (that controls how long
 * it is presented to the user). If the <code>markup</code> field is <code>true</code>,
 * the message is treated as HTML and used without any escaping.
 *
 * @since 5.3
 */
public class Alert implements Serializable
{
    private static final AtomicLong idSource = new AtomicLong(System.currentTimeMillis());

    /**
     * A unique id (unique within this JVM and execution), used to identify an alert (used primarily
     * when individually dismissing an alert).
     */
    public final long id = idSource.getAndIncrement();

    public final Duration duration;

    public final Severity severity;

    public final String message;
    
    /**
     * Defines whether the message will be treated as HTML or not.
     * @since 5.4
     */
    public final boolean markup;

    /**
     * Alert with default duration of {@link Duration#SINGLE} and default severity
     * of {@link Severity#INFO}.
     */
    public Alert(String message)
    {
        this(Severity.INFO, message);
    }

    /**
     * Alert with default duration of {@link Duration#SINGLE}.
     */
    public Alert(Severity severity, String message)
    {
        this(Duration.SINGLE, severity, message, false);
    }

    public Alert(Duration duration, Severity severity, String message)
    {
        this(duration, severity, message, false);
    }

    public Alert(Duration duration, Severity severity, String message, boolean markup)
    {
        assert duration != null;
        assert severity != null;
        assert InternalUtils.isNonBlank(message);

        this.duration = duration;
        this.severity = severity;
        this.message = message;
        this.markup = markup;
    }

    public String toString()
    {
        return String.format("Alert[%s %s %s %s]",
                duration.name(),
                severity.name(),
                message,
                markup);
    }

    public JSONObject toJSON()
    {
        JSONObject result = new JSONObject("message", message,
                "severity", severity.name().toLowerCase(), "markup", markup );

        if (duration == Duration.TRANSIENT)
        {
            result.put("transient", true);
        }

        if (duration.persistent)
        {
            result.put("id", id);
        }
        
        return result;
    }
}
