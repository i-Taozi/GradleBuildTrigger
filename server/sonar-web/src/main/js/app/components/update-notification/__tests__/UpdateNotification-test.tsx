/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { shallow } from 'enzyme';
import * as React from 'react';
import { getSystemUpgrades } from '../../../../api/system';
import { Alert } from '../../../../components/ui/Alert';
import DismissableAlert from '../../../../components/ui/DismissableAlert';
import { mockUpgrades } from '../../../../helpers/mocks/system-upgrades';
import { mockAppState, mockCurrentUser, mockLoggedInUser } from '../../../../helpers/testMocks';
import { waitAndUpdate } from '../../../../helpers/testUtils';
import { Permissions } from '../../../../types/permissions';
import { UpdateNotification } from '../UpdateNotification';

jest.mock('../../../../api/system', () => {
  const { mockUpgrades } = jest.requireActual('../../../../helpers/mocks/system-upgrades');
  return {
    getSystemUpgrades: jest
      .fn()
      .mockResolvedValue({ upgrades: [mockUpgrades({ version: '9.1' })], latestLTS: '8.9' })
  };
});

function formatDate(date: Date): string {
  return `${date.getFullYear()}-${date.getMonth() + 1}-${date.getDate()}`;
}

it('should render correctly', async () => {
  const wrapper = shallowRender({
    appState: mockAppState({ version: '9.0' }),
    currentUser: mockLoggedInUser({ permissions: { global: [Permissions.Admin] } })
  });
  await waitAndUpdate(wrapper);
  expect(wrapper).toMatchSnapshot('default');
  expect(wrapper.setProps({ currentUser: mockCurrentUser() })).toMatchSnapshot('anonymous user');
});

it('should not show prompt when not admin', async () => {
  //As anonymous
  const wrapper = shallowRender();
  await waitAndUpdate(wrapper);
  expect(wrapper.type()).toBeNull();

  // As non admin user
  wrapper.setProps({ currentUser: mockLoggedInUser() });
  await waitAndUpdate(wrapper);
  expect(wrapper.type()).toBeNull();
});

it('should not show prompt when no current version', async () => {
  const wrapper = shallowRender({ appState: mockAppState({ version: 'NOVERSION' }) });
  await waitAndUpdate(wrapper);
  expect(wrapper.type()).toBeNull();
});

it('should not show prompt when no upgrade', async () => {
  (getSystemUpgrades as jest.Mock).mockResolvedValueOnce({ upgrades: [], latestLTS: '8.9' });
  const wrapper = shallowRender({
    appState: mockAppState({ version: '9.1' }),
    currentUser: mockLoggedInUser({ permissions: { global: [Permissions.Admin] } })
  });
  await waitAndUpdate(wrapper);
  expect(wrapper.type()).toBeNull();
});

it('should show prompt when no lts date', async () => {
  (getSystemUpgrades as jest.Mock).mockResolvedValueOnce({
    upgrades: [mockUpgrades({ version: '8.9', releaseDate: 'INVALID' })],
    latestLTS: '8.9'
  });
  const wrapper = shallowRender({
    appState: mockAppState({ version: '8.1' }),
    currentUser: mockLoggedInUser({ permissions: { global: [Permissions.Admin] } })
  });
  await waitAndUpdate(wrapper);
  expect(wrapper.find(DismissableAlert).props().alertKey).toBe('previous_lts8.9');
  expect(wrapper.contains('admin_notification.update.previous_lts')).toBe(true);
});

it('should show prompt when minor upgrade', async () => {
  (getSystemUpgrades as jest.Mock).mockResolvedValueOnce({
    upgrades: [mockUpgrades({ version: '9.2' }), mockUpgrades({ version: '9.1' })],
    latestLTS: '8.9'
  });
  const wrapper = shallowRender({
    appState: mockAppState({ version: '9.1' }),
    currentUser: mockLoggedInUser({ permissions: { global: [Permissions.Admin] } })
  });
  await waitAndUpdate(wrapper);
  expect(wrapper.find(DismissableAlert).props().alertKey).toBe('new_minor_version9.2');
  expect(wrapper.contains('admin_notification.update.new_minor_version')).toBe(true);
});

it('should show prompt when patch upgrade', async () => {
  (getSystemUpgrades as jest.Mock).mockResolvedValueOnce({
    upgrades: [mockUpgrades({ version: '9.2' }), mockUpgrades({ version: '9.1.1' })],
    latestLTS: '8.9'
  });
  const wrapper = shallowRender({
    appState: mockAppState({ version: '9.1' }),
    currentUser: mockLoggedInUser({ permissions: { global: [Permissions.Admin] } })
  });
  await waitAndUpdate(wrapper);
  expect(wrapper.find(DismissableAlert).props().alertKey).toBe('new_patch9.2');
  expect(wrapper.contains('admin_notification.update.new_patch')).toBe(true);
});

it('should show prompt when lts upgrade', async () => {
  (getSystemUpgrades as jest.Mock).mockResolvedValueOnce({
    upgrades: [
      mockUpgrades({ version: '8.9', releaseDate: formatDate(new Date(Date.now())) }),
      mockUpgrades({ version: '9.2' }),
      mockUpgrades({ version: '9.1.1' })
    ],
    latestLTS: '8.9'
  });
  const wrapper = shallowRender({
    appState: mockAppState({ version: '8.8' }),
    currentUser: mockLoggedInUser({ permissions: { global: [Permissions.Admin] } })
  });
  await waitAndUpdate(wrapper);
  expect(wrapper.find(DismissableAlert).props().alertKey).toBe('pre_lts8.9');
  expect(wrapper.contains('admin_notification.update.pre_lts')).toBe(true);
});

it('should show prompt when lts upgrade is more than 6 month', async () => {
  const ltsDate = new Date(Date.now());
  ltsDate.setMonth(ltsDate.getMonth() - 7);
  (getSystemUpgrades as jest.Mock).mockResolvedValueOnce({
    upgrades: [
      mockUpgrades({ version: '8.9', releaseDate: formatDate(ltsDate) }),
      mockUpgrades({ version: '9.2' }),
      mockUpgrades({ version: '9.1.1' })
    ],
    latestLTS: '8.9'
  });
  const wrapper = shallowRender({
    appState: mockAppState({ version: '8.8' }),
    currentUser: mockLoggedInUser({ permissions: { global: [Permissions.Admin] } })
  });
  await waitAndUpdate(wrapper);
  expect(wrapper.find(DismissableAlert).props().alertKey).toBe('previous_lts8.9');
  expect(wrapper.contains('admin_notification.update.previous_lts')).toBe(true);
});

it('should show correct alert when not dismissable', async () => {
  (getSystemUpgrades as jest.Mock).mockResolvedValueOnce({
    upgrades: [
      mockUpgrades({ version: '8.9', releaseDate: formatDate(new Date(Date.now())) }),
      mockUpgrades({ version: '9.2' }),
      mockUpgrades({ version: '9.1.1' })
    ],
    latestLTS: '8.9'
  });
  const wrapper = shallowRender({
    appState: mockAppState({ version: '8.8' }),
    currentUser: mockLoggedInUser({ permissions: { global: [Permissions.Admin] } })
  });
  await waitAndUpdate(wrapper);
  expect(wrapper.find(DismissableAlert).type).toBeDefined();
  wrapper.setProps({ dismissable: false });
  expect(wrapper.find(Alert).type).toBeDefined();
});

function shallowRender(props: Partial<UpdateNotification['props']> = {}) {
  return shallow(
    <UpdateNotification
      dismissable={true}
      appState={mockAppState()}
      currentUser={mockCurrentUser()}
      {...props}
    />
  );
}
