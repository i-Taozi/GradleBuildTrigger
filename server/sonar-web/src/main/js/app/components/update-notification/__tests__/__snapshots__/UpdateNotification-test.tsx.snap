// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly: anonymous user 1`] = `""`;

exports[`should render correctly: default 1`] = `
<DismissableAlert
  alertKey="new_minor_version9.1"
  className="promote-update-notification it__upgrade-prompt-new_minor_version"
  variant="info"
>
  admin_notification.update.new_minor_version
  <SystemUpgradeButton
    latestLTS="8.9"
    systemUpgrades={
      Array [
        Object {
          "changeLogUrl": "changelogurl",
          "description": "Version 5.6.7 description",
          "downloadUrl": "downloadurl",
          "releaseDate": "2017-03-01",
          "version": "9.1",
        },
      ]
    }
    updateUseCase="new_minor_version"
  />
</DismissableAlert>
`;
